### PR TITLE
Issue 428

### PR DESCRIPTION
**Pull request by skodapetr**
This pull request was created to address issue #428.

The proposed solution is to move tokens management into ApiConnection class. The idea is that the ApiConnection keeps track of token requests and store them. This allows reusing the token when necessary, for example for logout operation. To reflect this fact in the API of ApiConnection  a new method *getOrFetchToken* was introduced. I'm not sure about the name, as we can keep *fetchToken* and hide the fact that it may not always cause fetch operation. I prefer a more explicit solution although they are more verbose, that is personal preference.

As tokens are managed, and cached, in ApiConnection the WbEditingAction class does not store the *CSRF* token and instead, it gets it from the ApiConnection.

The tests were also changed to reflect a fact, that *CSRF* can be obtained only after login. This seems to be original intention, based on the resource test file names but was not implemented in tests.

It seems that using the *CSRF* token works with both instances I tested with (1.32.0, 1.34.0-wmf.17), so the change should be backward compatible. But the older version API complain that the token is unknown.
```
API warning [main]: Unrecognized parameter: token.
```

### CODE REVIEW

File: wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/BasicApiConnection.java [hunk 1]
```diff
@@ -57,14 +57,17 @@ public BasicApiConnection(String apiBaseUrl) {
...
 	 */
 	@JsonCreator
 	private BasicApiConnection(
 			@JsonProperty("baseUrl") String apiBaseUrl,
 			@JsonProperty("cookies") Map<String, String> cookies,
 			@JsonProperty("username") String username,
-			@JsonProperty("loggedIn") boolean loggedIn) {
-		super(apiBaseUrl, cookies, username, loggedIn);
+			@JsonProperty("loggedIn") boolean loggedIn,
+			@JsonProperty("cookies") Map<String, String> tokens) {
```
Comment (wetneb) [inline wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/BasicApiConnection.java]: There seems to be a mismatch between the serialization key `"cookies"` and the variable name `tokens`.

Comment (skodapetr) [reply wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/BasicApiConnection.java]: Thanks for pointing that out, should be fixed by now.

File: wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/WbEditingAction.java [hunk 2]
```diff
@@ -788,40 +782,12 @@ private EntityDocument parseJsonResponse(JsonNode entityNode) throws IOException
...
-	 * Obtains and sets a new CSRF token, whether or not there is already a
-	 * token set right now.
-	 */
-	private void refreshCsrfToken() {
-
-		this.csrfToken = fetchCsrfToken();
+		return connection.getOrFetchToken("csrf");
 		// TODO if this is null, we could try to recover here:
 		// (1) Check if we are still logged in; maybe log in again
 		// (2) If there is another error, maybe just run the operation again
```
Comment (Tpt) [inline wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/WbEditingAction.java]: This comment does not look relevant anymore. I believe it should be moved to the `ApiConnection.getOrFetchToken` method. The `getCsrfToken` might also be removed now that it is just composed of one method call.

File: wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/ApiConnection.java [hunk 3]
```diff
@@ -335,6 +345,7 @@ public void logout() throws IOException {
 			Map<String, String> params = new HashMap<>();
 			params.put("action", "logout");
 			params.put("format", "json"); // reduce the output
+			params.put("token", fetchToken("csrf"));
```
Comment (Tpt) [inline wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/ApiConnection.java]: Maybe`getOrFetchToken` if the csrf token has been already loaded?

Comment (skodapetr) [reply wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/ApiConnection.java]: Yes, I probably missed that as I originally tested with this code.
One more thing to consider might the re-fresh of CSRF certificate as in ```WBEditAction```, as I'm not sure whether the close may fail with "older"? CSRF certificate.

**Comment by skodapetr** [comment]
@Tpt Thanks for the comments.
I updated the pull request to reflect the comments.

I understand that every programmer has their own preferences, so please feel free to update anything to make it match the code style and chosen architecture of this repository.

File: wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/WbEditingAction.java (hunk 2 above)
Comment (skodapetr) [reply wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/WbEditingAction.java]: Should be resolved.

**Comment by Tpt** [comment]
Thank you very much!

**Pull request merged**